tests/run/errorhandling/Test.scala

```diff
@@ -0,0 +1,78 @@
+import dotty.util.*
+
+/** boundary/break as a replacement for non-local returns */
+def indexOf[T](xs: List[T], elem: T): Int =
+  boundary:
+    for (x, i) <- xs.zipWithIndex do
+      if x == elem then break(i)
```

Comment:
It really feels like we're closer to having Python generator syntax, which is fantastic! Would love to see a strawman for this if it is actually possible with the existing Scala 3 capabilities.

In Python you can do things like this, which read naturally and return a generator, an equivalent to a Scala/Java Iterator:

```python
def example():
  yield "We'll give you all the numbers divisible by 3 or 2, with those by 3 as priority"
  for i in range(1, 1000):
    if i % 3 == 0:
      yield ("%d is divisible by 3" % i)
    elif i % 2 == 0:
      yield ("%d is even" % i)
```

![image](https://user-images.githubusercontent.com/2464813/210676571-9847d2de-9b55-4e66-8ac6-89ada877a60c.png)

The `yield`s can live anywhere, and you can easily `return` or `break` out.

And of course you can use this generator in another for-loop, or even manually consume it. They terminate through an exception called `StopIteration`:
![image](https://user-images.githubusercontent.com/2464813/210677008-7344569a-2965-402f-9617-f57716df473f.png)

However, I am not sure if it is possible to at all solve, but I have not seen clean-up/finalization in generators. This is where libraries like FS2 really shine because you have guaranteed clean-up of resources. I'm willing to put up with code that is more rigid in order to get that guarantee.

Reply:
> but if there is a way to achieve the best of both worlds it would be really fantastic.

@ScalaWilliam you mean like this? :)

<details>

```scala
//> using lib "com.github.rssh::dotty-cps-async::0.9.14"
//> using lib "co.fs2::fs2-core::3.4.0-56-9e373bd-SNAPSHOT"
//> using repository "https://s01.oss.sonatype.org/content/repositories/snapshots/"
//> using option "-Ykind-projector"

import cats.effect.*
import fs2.*
import cps.*

given [O]: CpsMonad[Pull[IO, O, *]] with CpsMonadInstanceContext[Pull[IO, O, *]] with
  def pure[A](a: A): Pull[IO, O, A] = Pull.pure(a)
  def map[A, B](fa: Pull[IO, O, A])(f: A => B): Pull[IO, O, B] = fa.map(f)
  def flatMap[A, B](fa: Pull[IO, O, A])(f: A => Pull[IO, O, B]): Pull[IO, O, B] = fa.flatMap(f)

def generator: Pull[IO, String, Unit] =
  async[Pull[IO, String, *]] {
    await(
      Pull.acquire[IO, Unit](
        IO.println("acquired resource"),
        (_, _) => IO.println("released resource")
      ): Pull[IO, String, Unit]
    )

    (1 to 20).map { i =>
      if i % 3 == 0 then
        await(Pull.output1[IO, String](s"$i is divisible by 3"))
      if i % 2 == 0 then
        await(Pull.output1[IO, String](s"$i is even"))
    }
  }

@main def main =
  import cats.effect.unsafe.implicits.global
  generator.stream.foreach(IO.println(_)).compile.drain.unsafeRunSync()
```

</details>

```
acquired resource
2 is even
3 is divisible by 3
4 is even
6 is divisible by 3
6 is even
8 is even
9 is divisible by 3
10 is even
12 is divisible by 3
12 is even
14 is even
15 is divisible by 3
16 is even
18 is divisible by 3
18 is even
20 is even
released resource
```

I'm a broken record, but CPS is really really awesome!

Reply:
I don't disagree :) mixing resources with generators is somewhat "advanced", because the resource needs to stay alive for the lifetime of the generator, and the lifetime of the generator is determined by some downstream consumer of it.

That is to say: the generator itself _is_ the "envelope", since we don't know what goes "inside" until the user actually consumes it (and does whatever).

So in the example above, bringing a resource into the generator process keeps it alive and makes it available for the entire lifetime of the generator (as determined by the consumer).

Reply:
@armanbilge that is impressive, I did not expect it would even be possible in fs2. Even experimenting with vars inside this just works.

Are you able to also do an fs2 Pipe that would 'await' for an input, and conditionally produce outputs? Possibly allocating a resource in the middle as well like you had declared.

To make it very simple, is it possible to write the equivalent of `.filter(_ % 3 == 0)` in this style as well?

Reply:
@ScalaWilliam Thanks!

> Even experimenting with vars inside this just works.

It might, but it would be safer to use `await(Pull.eval(IO.ref(...)))` to make mutable vars, and update them with `await(Pull.eval(ref.set(...)))` 😄 

---

> Are you able to also do an fs2 Pipe that would 'await' for an input, and conditionally produce outputs?

Great question, we've been discussing this topic on the Typelevel Discord. So first of all, you can definitely do this with the `Stream` monad:

<details>

```scala
//> using lib "com.github.rssh::dotty-cps-async::0.9.14"
//> using lib "co.fs2::fs2-core::3.4.0-56-9e373bd-SNAPSHOT"
//> using repository "https://s01.oss.sonatype.org/content/repositories/snapshots/"
//> using option "-Ykind-projector"

import cats.effect.*
import fs2.*
import cps.*

given CpsMonad[Stream[IO, *]] with CpsMonadInstanceContext[Stream[IO, *]] with
  def pure[A](a: A): Stream[IO, A] = Stream.emit(a)
  def map[A, B](fa: Stream[IO, A])(f: A => B): Stream[IO, B] = fa.map(f)
  def flatMap[A, B](fa: Stream[IO, A])(f: A => Stream[IO, B]): Stream[IO, B] = fa.flatMap(f)

def pipe(in: Stream[IO, Int]): Stream[IO, Int] =
  async[Stream[IO, *]] {
    val i = await(in)
    if i % 3 == 0 then
      await(Stream.emit[IO, Int](i))
    else
      await(Stream.empty: Stream[IO, Int])
  }

@main def main =
  import cats.effect.unsafe.implicits.global
  Stream.emits(1 to 20).through(pipe).foreach(IO.println(_)).compile.drain.unsafeRunSync()
```

</details>

However it's not yet possible to do this in the `Pull` monad, which is ideal for writing "generators".

There are a few ways to transform a `Pull[IO, Output, State]` datatype. My original example relied on `flatMap`, which gives you access to the current `State`.
```scala
def flatMap[A, B](pullA: Pull[IO, Out, A])(f: A => Pull[IO, Out, B]): Pull[IO, Out, B]
```
Then, CPS transforms `await` to `flatMap`:
```scala
val a = await(pullA)
await(f(a))
```

`Pull` also offers a method `flatMapOutput`, which gives you access to the current `Out`.
```scala
def flatMapOutput(in: Pull[IO, Out, Unit])(f: Out => Pull[IO, Out, Unit]): Pull[IO, Out, Unit]
```
It would be interesting if we could add e.g. an `awaitNext(...)` syntax that transforms to `flatMapOutput` under-the-hood:
```scala
val i = awaitNext(in)
if i % 3 == 0 then
  await(Pull.output1(i))
```

I don't know enough about the inner-workings of CPS to know if this is feasible. But it seems like it could be :)